File: node-server/api/send-email.js
import nodemailer from "nodemailer";

export default async function handler(req, res) {
  if (req.method !== "POST") {
    return res.status(405).json({ success: false, message: "Method not allowed" });
  }

  try {
    const { to, subject, text, html } = req.body;

    if (!to || !subject || (!text && !html)) {
      return res.status(400).json({ success: false, message: "Missing required fields" });
    }

    const transporter = nodemailer.createTransport({
      service: "gmail",
      auth: {
        user: process.env.GMAIL_USER,
        pass: process.env.GMAIL_APP_PASSWORD,
      },
    });

    await transporter.sendMail({
      from: `"Zendo" <${process.env.GMAIL_USER}>`,
      to,
      subject,
      text,
      html,
    });

    return res.status(200).json({ success: true, message: "Email sent successfully" });
  } catch (error) {
    console.error("Email error:", error);
    return res.status(500).json({ success: false, message: "Failed to send email" });
  }
}

File: go-server/api/send-email.go
package handler

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"sync"

	gomail "gopkg.in/mail.v2"
)

var (
	dialer *gomail.Dialer
	sender gomail.SendCloser
	once   sync.Once
)

type EmailRequest struct {
	To      string `json:"to"`
	Subject string `json:"subject"`
	HTML    string `json:"html,omitempty"`
	From    string `json:"from"`
}

type JsonResponse struct {
	Success bool   `json:"success"`
	Message string `json:"message"`
}

func writeJSON(w http.ResponseWriter, status int, payload JsonResponse) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(payload)
}

func Handler(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeJSON(w, http.StatusMethodNotAllowed, JsonResponse{false, "Method not allowed"})
		return
	}

	var req EmailRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSON(w, http.StatusBadRequest, JsonResponse{false, "Invalid JSON body"})
		return
	}

	if req.To == "" || req.Subject == "" || req.HTML == "" {
		writeJSON(w, http.StatusBadRequest, JsonResponse{false, "Missing required fields"})
		return
	}

	if err := SendEmail(req.From, req.To, req.Subject, req.HTML); err != nil {
		writeJSON(w, http.StatusInternalServerError, JsonResponse{false, "Failed to send email: " + err.Error()})
		return
	}

	log.Println("Email Sent successfully to " + req.To + " from " + req.From)
	writeJSON(w, http.StatusOK, JsonResponse{true, "Email sent successfully"})
}

func initDialer() {
	dialer = gomail.NewDialer("smtp.gmail.com", 465, os.Getenv("GMAIL_USER"), os.Getenv("GMAIL_APP_PASSWORD"))
	var err error
	sender, err = dialer.Dial()
	if err != nil {
		log.Fatalf("Failed to dial SMTP: %v", err)
	}
}

func SendEmail(from, to, subject, body string) error {
	once.Do(initDialer)

	m := gomail.NewMessage()
	m.SetHeader("From", fmt.Sprintf("%s <%s>", from, os.Getenv("GMAIL_USER")))
	m.SetHeader("To", to)
	m.SetHeader("Subject", subject)
	m.SetBody("text/html", body)

	return gomail.Send(sender, m)
}
